### PR TITLE
[in_app_purchase] Fix finishing purchases upon payment dialog cancel…

### PR DESCRIPTION
**Pull request by dennis-tra**
## Description

This PR fixes finishing purchases upon payment dialog cancellation on iOS.

When the user cancels the payment dialog there is no `transactionIdentifier` which means the transaction cannot be finished in the current implementation and therefore not initiated again.

Duplicate of:
- https://github.com/flutter/plugins/pull/3074

## Related Issues

- https://github.com/flutter/flutter/issues/66253
- https://github.com/flutter/flutter/issues/66886

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Comment by dennis-tra** [comment]
@googlebot I signed it!

File: packages/in_app_purchase/ios/Classes/InAppPurchasePlugin.m [hunk 1]
```diff
@@ -197,19 +197,27 @@ - (void)addPayment:(FlutterMethodCall *)call result:(FlutterResult)result {
...
+  NSDictionary *paymentMap = (NSDictionary *)call.arguments;
+  NSString *transactionIdentifier = [paymentMap objectForKey:@"transactionIdentifier"];
+  NSString *productIdentifier = [paymentMap objectForKey:@"productIdentifier"];
 
   NSArray<SKPaymentTransaction *> *pendingTransactions =
       [self.paymentQueueHandler getUnfinishedTransactions];
 
   for (SKPaymentTransaction *transaction in pendingTransactions) {
-    if ([transaction.transactionIdentifier isEqualToString:transactionIdentifier]) {
+    // If the user cancels the purchase dialog we won't have a transactionIdentifier. So if it is
```
Comment (LHLL) [inline packages/in_app_purchase/ios/Classes/InAppPurchasePlugin.m:214]: I think these two words break the line. Same below.

File: packages/in_app_purchase/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart [hunk 2]
```diff
@@ -103,9 +103,11 @@ class SKPaymentQueueWrapper {
   /// finishTransaction:]`](https://developer.apple.com/documentation/storekit/skpaymentqueue/1506003-finishtransaction?language=objc).
   Future<void> finishTransaction(
       SKPaymentTransactionWrapper transaction) async {
+    Map requestMap = transaction.toFinishMap();
```
Comment (LHLL) [inline packages/in_app_purchase/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart:106]: Prefer lightweight generic. Consider adding a type for the `Map`. For example: Map<String, String>

**Review by LHLL — COMMENTED** [review]
Thanks for fixing this!

File: packages/in_app_purchase/ios/Classes/InAppPurchasePlugin.m (hunk 1 above)
Comment (dennis-tra) [reply packages/in_app_purchase/ios/Classes/InAppPurchasePlugin.m]: I reformatted the comment

File: packages/in_app_purchase/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart (hunk 2 above)
Comment (dennis-tra) [reply packages/in_app_purchase/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart]: Done

**Review by LHLL — APPROVED** [review]
LGTM! Thanks.

**Review by cyanglaz — CHANGES_REQUESTED** [review]
Looks like the version needs to be updated from pubspec.yaml

**Comment by luisdhm** [comment]
Still getting an error after completePurchase method in iOS 14.0.1.
PlatformException(storekit_invalid_argument, Argument type of finishTransaction is not a Dictionary, null, null)

**Comment by fbardecio** [comment]
Hi! I'm experiencing the same error PlatformException(storekit_invalid_argument, Argument type of finishTransaction is not a Dictionary, null, null)

**Comment by DangKhoi1997** [comment]
Updated to version 0.3.4+12 but it still not pop up purchase dialog :(. Anyone has same error ??

**Comment by LHLL** [comment]
> Updated to version 0.3.4+12 but it still not pop up purchase dialog :(. Anyone has same error ??
@DangKhoi1997 I think dialog not showing up is a different issue and not related to this PR. Apple Pay UI/dialog is not showing up is related to not finishing transactions. Can you try something similar to following code to see if your problem is solved:
```
class PurchaseService{
    final _connection = InAppPurchaseConnection();

    const PurchaseService(){
          _connection.purchaseUpdatedStream.listen((detail) {
               ...
              // Verify the receipt for succeeded transactions and handle the error for the failed
              // transaction here.
              _connection.completePurchase(detail);
          });    
    }
}
```

**Comment by DangKhoi1997** [comment]
> you
U cant pass the detail(List<PurchaseDetails>) into completePurchase because Future<BillingResultWrapper> completePurchase(PurchaseDetails purchase,{String developerPayload});
Any workaround to track or clear the transactions ?

**Comment by DangKhoi1997** [comment]
@LHLL  U cant pass the detail(List<PurchaseDetails>) into completePurchase because Future<BillingResultWrapper> completePurchase(PurchaseDetails purchase,{String developerPayload});
Any workaround to track or clear the transactions ?

**Comment by DangKhoi1997** [comment]
After restarting ip the issue gone but exist when hitting cancel button 3-4 times.

**Comment by daoc13** [comment]
Updated to latest version 0.3.4+12 but we are still not able to buy again whenever we cancel the purchase dialogue. The button works fine if we complete the purchase flow and I am able to buy many time if the purchase flow ends correct, but whenever I cancel and try to tap again the button it won’t open the purchase dialogue. Do we need to add any additional code to our implementation?

**Comment by hai199580** [comment]
> Updated to latest version 0.3.4+12 but we are still not able to buy again whenever we cancel the purchase dialogue. The button works fine if we complete the purchase flow and I am able to buy many time if the purchase flow ends correct, but whenever I cancel and try to tap again the button it won’t open the purchase dialogue. Do we need to add any additional code to our implementation?

@daoc13 do you find any solution for this issue :((

**Comment by daoc13** [comment]
> > Updated to latest version 0.3.4+12 but we are still not able to buy again whenever we cancel the purchase dialogue. The button works fine if we complete the purchase flow and I am able to buy many time if the purchase flow ends correct, but whenever I cancel and try to tap again the button it won’t open the purchase dialogue. Do we need to add any additional code to our implementation?
> 
> @daoc13 do you find any solution for this issue :((

@hai199580 I did not find a solution for this issue. I recommend you to use Revenue Cat, we implemented it in 4 hours and it works fine. I will keep using Revenue Cat until this issue is solved. Hope it helps!

**Pull request merged**